### ISSUE TITLE
Combining Yolov8 with lstm for  spatio-temporal action recognition

### ISSUE THREAD

**Issue by dvskabangira** [ultralytics/ultralytics#8846]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

i would like to combine yolov8 with lstm for spatio-temporal action, can this be done? Thank you for the feedback

**Comment by glenn-jocher** [comment]
@dvskabangira certainly! Combining YOLOv8 with LSTM for spatio-temporal action recognition is a promising approach. While YOLOv8 excels in spatial detection, LSTM can help leverage temporal information across frames. Here's a basic outline to get you started:

1. Use YOLOv8 to detect objects of interest in each frame.
2. Extract the bounding box coordinates or feature vectors of these detections.
3. Feed this sequential data into an LSTM network to model temporal dependencies.

Here's a simplified code snippet to illustrate the concept:

```python
from ultralytics import YOLO
import torch.nn as nn

# Load YOLOv8 model
model = YOLO('yolov8n.pt')

# Define your LSTM model
class ActionLSTM(nn.Module):
    def __init__(self, input_size, hidden_size, num_layers, num_classes):
        super(ActionLSTM, self).__init__()
        self.lstm = nn.LSTM(input_size, hidden_size, num_layers, batch_first=True)
        self.fc = nn.Linear(hidden_size, num_classes)
    
    def forward(self, x):
        # Forward propagate LSTM
        out, _ = self.lstm(x)
        # Decode the hidden state of the last time step
        out = self.fc(out[:, -1, :])
        return out

# Example LSTM parameters
input_size = 2048  # Depends on YOLOv8 output feature size
hidden_size = 512
num_layers = 2
num_classes = 10  # Number of actions to recognize

action_lstm = ActionLSTM(input_size, hidden_size, num_layers, num_classes)

# Example: Process a sequence of frames with YOLOv8 and LSTM
# frames = [frame1, frame2, ..., frameN]  # Your video frames here
# yolo_outputs = [model(frame) for frame in frames]  # Detect with YOLOv8
# lstm_inputs = prepare_lstm_inputs(yolo_outputs)  # Convert YOLOv8 outputs to LSTM inputs
# lstm_outputs = action_lstm(lstm_inputs)  # Predict actions with LSTM
```

Remember, the key is in how you prepare the LSTM inputs (`prepare_lstm_inputs` function in the snippet) from YOLOv8 outputs. This might involve selecting specific features or transforming bounding box coordinates.

Feel free to dive deeper into the implementation details based on your project's specific needs. Good luck! 🚀

**Comment by BROHAO1** [comment]
@glenn-jocher super useful explanation, thanks! Could you explain further about which feature from YOLOv8 outputs should be feed to LSTM,  and how to transform related coordinates.

**Comment by glenn-jocher** [comment]
@BROHAO1 I'm glad you found the explanation helpful! For feeding YOLOv8 outputs to an LSTM, you typically want to use the bounding box coordinates (x, y, width, height) and class probabilities. These features encapsulate both spatial and categorical information about the detected objects.

Here’s a quick way to transform these coordinates for LSTM processing:

```python
# Assuming 'results' is the output from YOLOv8
boxes = results.boxes.xywh  # Get bounding boxes in xywh format
classes = results.boxes.cls  # Get class IDs
confidences = results.boxes.conf  # Get confidence scores

# Normalize coordinates by image dimensions (e.g., 640x640)
normalized_boxes = boxes / 640

# Prepare LSTM input as a concatenation of normalized coordinates, class, and confidence
lstm_input = torch.cat((normalized_boxes, classes.unsqueeze(1), confidences.unsqueeze(1)), dim=1)
```

This format keeps the spatial and class information in a form that's easy for the LSTM to process, capturing both where the object is and what it is likely to be. Hope this helps! 😊

**Comment by SauirbekA** [comment]
Hi there, @glenn-jocher , can i use key point detection for this task, if so, would you please suggest any ideas?)

**Comment by glenn-jocher** [comment]
Hi @SauirbekA! Absolutely, you can use keypoint detection for tasks that require understanding of object poses or specific parts. YOLOv8 can be extended with keypoint detection by training it on a dataset annotated with keypoints (like human joints for pose estimation).

Here’s a simple example to get you started:

```python
from ultralytics import YOLO

# Load a model trained on keypoint detection
model = YOLO('yolov8n-pose.pt')

# Run prediction
results = model('path/to/image.jpg')

# Access keypoints
keypoints = results.keypoints
```

This will give you the keypoints for each detected object. You can then use these for further analysis or visualization. Happy coding! 😊

**Comment by EbranurMeric** [comment]
Hi,  @glenn-jocher How to estimate pose and motion in sports using YOLO and LSTM. Also labeling sports videos as true or false.Would you please suggest any ideas?

**Comment by glenn-jocher** [comment]
You can treat this as an action‑recognition pipeline: use a pose model (e.g., `yolo11n-pose.pt`) to extract keypoints per frame, stack them over time, and train an LSTM that outputs a binary label for each clip (“true” = target motion present, “false” = not present), for example:

```python
from ultralytics import YOLO
import torch.nn as nn
import torch

pose_model = YOLO("yolo11n-pose.pt")

def extract_sequence(video_path):
    seq = []
    for r in pose_model(video_path, stream=True):
        if r.keypoints is None:
            continue
        k = r.keypoints.xyn[0, :, :2].reshape(-1)  # first person, normalized x,y
        seq.append(k)
    return torch.stack(seq)  # [T, K*2]

class ClipLSTM(nn.Module):
    def __init__(self, input_size, hidden_size=256):
        super().__init__()
        self.lstm = nn.LSTM(input_size, hidden_size, batch_first=True)
        self.fc = nn.Linear(hidden_size, 2)  # true / false

    def forward(self, x):
        y, _ = self.lstm(x)
        return self.fc(y[:, -1])

# Example usage:
seq = extract_sequence("sports_clip.mp4").unsqueeze(0)  # [1, T, K*2]
model = ClipLSTM(input_size=seq.shape[-1])
logits = model(seq)
```

For a high‑level overview of this kind of human activity recognition pipeline, see the [Action Recognition (HAR) explanation](https://www.ultralytics.com/glossary/action-recognition).
